app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java

```diff
@@ -667,4 +667,43 @@ public void generateAndSetPagePolicies(Application application, PageDTO page) {
                 });
     }
 
+    @Override
+    public Mono<Application> reorderPage(String applicationId, String pageId, Integer order) {
+        return applicationService.findById(applicationId, MANAGE_APPLICATIONS)
+                .switchIfEmpty(Mono.error(new AppsmithException(AppsmithError.ACL_NO_RESOURCE_FOUND, FieldName.APPLICATION, applicationId)))
+                .flatMap(application -> {
+                    // Update the order in unpublished pages here, since this should only ever happen in edit mode.
+                    final List<ApplicationPage> pages = application.getPages();
+
+                    ApplicationPage foundPage = null;
+                    for (final ApplicationPage page : pages) {
+                        if (pageId.equals(page.getId())) {
+                            foundPage = page;
+                        }
+                    }
+                    //Set the order for the pages which has null values. At present we dont store the order and hence we need to update the values
+                    for (int i = 1; i <= pages.size(); i++) {
+                        ApplicationPage page = pages.get(i - 1);
+                        if (page.getOrder() == null) {
+                            page.setOrder(i);
+                        }
+                    }
+
+                    if(foundPage != null){
+                        //Increment the order of subsequent pages
+                        for (final ApplicationPage page : pages) {
+                            if (page.getOrder() < foundPage.getOrder()) {
+                                page.setOrder(page.getOrder()+1);
+                            }
+                        }
+
+                        //set the selected page order to 0
+                        foundPage.setOrder(0);
```

Comment:
We should set the given `order` for the page we found no?

```suggestion
                        foundPage.setOrder(order);
```

Reply:
This is bad, i missed this. I will change this.